openedx_tagging/core/tagging/models/base.py

```diff
@@ -342,6 +342,117 @@ def get_filtered_tags(
 
         return tag_set.order_by("value", "id")
 
+    def add_tag(
+        self,
+        tag_value: str,
+        parent_tag_value: str | None = None,
+        external_id: str | None = None
+    ) -> Tag:
+        """
+        Add new Tag to Taxonomy. If an existing Tag with the `tag_value` already
+        exists in the Taxonomy, an exception is raised, otherwise the newly
+        created Tag is returned
+        """
+        self.check_casted()
+
+        if self.allow_free_text:
+            raise ValueError(
+                "add_tag() doesn't work for free text taxonomies. They don't use Tag instances."
+            )
+
+        if self.system_defined:
+            raise ValueError(
+                "add_tag() doesn't work for system defined taxonomies. They cannot be modified."
+            )
+
+        if self.tag_set.filter(value__iexact=tag_value).exists():
+            raise ValueError(f"Tag with value '{tag_value}' already exists for taxonomy.")
+
+        parent = None
+        if parent_tag_value:
+            # Check if parent tag is valid
+            if not Tag.objects.filter(value__iexact=parent_tag_value).exists():
+                raise ValueError("Invalid `parent_tag_value` provided")
```

Comment:
I don't understand why we even need these lines ^^. Isn't the check below, which raises `Tag.DoesNotExist` sufficient? This just does an extra database query for no reason. We don't really care if the value exists already but in a different taxonomy - the check below already accounts for "value exists in another taxonomy" and "value doesn't exist at all"

Reply:
It is indeed sufficient, I just thought it adds a bit more granularity and description to the errors for the end user. However, you're right, it seems that the extra DB query for this might not be worthwhile. I've removed all of them and updated some test cases.